### PR TITLE
feat: Integrate with opensea API to fetch collectible

### LINKED ISSUE #263: implement basic opensea integration
related to https://github.com/status-im/status-desktop/issues/2337

goal: integrate with https://docs.opensea.io/reference/api-overview so that we can get the NFTs/collectibles

objective 1: return list of collectibles

typing `/assets 0x12345` should return a list of collectibles associated to that account 0x12345

example possible output:

```
[
  {name, image_thumbnail, image, collection_name, collection_address}
]
```

objective 2: like object 1 but should return organized by collectibles

yping `/assets 0x12345` should return a list of collectibles associated to that account 0x12345

example possible output:

```
{
  <collectible_name>:  [
    {name, image_thumbnail, image, collection_address}
  ]
 <collectible_name2>:  [
    {name, image_thumbnail, image, collection_address},
    {name, image_thumbnail, image, collection_address}
  ]

}
```

note: can skip database / database is optional in this task
note: can use a json response in the TUI

### PR DESCRIPTION
**Pull request by alaibe**
fixes #263 

image

@michaelsbradleyjr @emizzle if you could help on the review ;) I am new to the repo

### CODE REVIEW

File: status/api/opensea.nim [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    imageUrl: jsonMsg{"image_url"}.getStr,
+    address: jsonMsg{"asset_contract"}{"address"}.getStr,
+    collectionName: jsonMsg{"collection"}{"name"}.getStr
+  )
+
+proc queryAssets(client: HttpClient, owner: Address, offset: int): seq[Asset] =
+  var content = ""
+  var assets: seq[Asset] = @[]
+  try:
+    content = client.getContent(fmt("https://api.opensea.io/api/v1/assets?owner={owner}&offset={offset}&limit=50"))
```
Comment (michaelsbradleyjr) [inline status/api/opensea.nim:30]: Should become a chronos `fetch`.

File: status/api/opensea.nim [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      if len(tmpAssets) < 50:
+        break
+  finally:
+    client.close()
+
+  return assets
+
+proc getOpenseaAssets*(self: StatusObject, owner: Address): OpenseaAssetsResult =
+  try:
+    let assets = queryAllAssets(owner)
```
Comment (michaelsbradleyjr) [inline status/api/opensea.nim:61]:
```suggestion
    let assets = await queryAllAssets(owner)
```

**Review by michaelsbradleyjr — CHANGES_REQUESTED** [review]
Really nice work, especially considering how new you are to the codebase, and Nim in general! I hope it will come as a compliment when I say I'm not surprised.

Let me know if you have any questions re: my requested changes; my comments probably leave some questions unanswered and probably assume knowledge/understanding of things you aren't acquainted with yet.

File: status/api/opensea.nim [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  var assets: seq[Asset] = @[]
+  try:
+    content = client.getContent(fmt("https://api.opensea.io/api/v1/assets?owner={owner}&offset={offset}&limit=50"))
+  except:
+    return assets
+
+  let doc = content.parseJson()
+
+  for json in doc["assets"]:
+    assets.add(toAsset(json))
```
[the remark this answers is not in the document]
Comment (emizzle) [reply status/api/opensea.nim:37]: To de/serialize an `Asset` correctly, we'll need to provide some metadata on the `Asset` type like so:
```nim
type
  Asset* = object
    name*: string
    imageThumbnailUrl* {.serializedFieldName("image_thumbnail_url").}: string
    imageUrl* {.serializedFieldName("image_url").}: string
    contract* {.serializedFieldName("asset_contract").}: AssetContract
    collection*: Collection
  AssetContract* = object
    address*: Address
  Collection* = object
    name*: string
```

File: status/api/opensea.nim [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+proc toAsset(jsonMsg: JsonNode): Asset =
+  return Asset(
+    name: jsonMsg{"name"}.getStr,
+    imageThumbnailUrl: jsonMsg{"image_thumbnail_url"}.getStr,
+    imageUrl: jsonMsg{"image_url"}.getStr,
+    address: jsonMsg{"asset_contract"}{"address"}.getStr,
+    collectionName: jsonMsg{"collection"}{"name"}.getStr
+  )
+
+proc queryAssets(client: HttpClient, owner: Address, offset: int): seq[Asset] =
```
[the remark this answers is not in the document]
Comment (emizzle) [reply status/api/opensea.nim:26]:
```suggestion
type
  QueryAssetsResult = Result[seq[Asset], string]
 
proc queryAssets(client: HttpClient, owner: Address, offset: int): QueryAssetsResult =
```

File: status/api/opensea.nim [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    return assets
+
+  let doc = content.parseJson()
+
+  for json in doc["assets"]:
+    assets.add(toAsset(json))
+
+  return assets
+
+proc queryAllAssets(owner: Address): seq[Asset] =
```
[the remark this answers is not in the document]
Comment (emizzle) [reply status/api/opensea.nim:41]: And it likely should be async as well. Maybe @michaelsbradleyjr  has more to input on that.

**Review by emizzle — CHANGES_REQUESTED** [review]
Overall, amazing work @alaibe! Great to have you contributing to nim-status! There are a couple of things that need to be updated, namely exception handling and swapping to nim-chronos' `AsyncHttpClient`. But I'd like to echo @michaelsbradleyjr's comments and say that I am impressed by this work considering how this was your first look at the repo!

File: status/api/opensea.nim [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  var offset = 0
+  var assets: seq[Asset] = @[]
+  let client = newHttpClient()
```
[the remark this answers is not in the document]
Comment (emizzle) [reply status/api/opensea.nim:44]: Going along with @michaelsbradleyjr 's suggestion and the Result return type changes I mentioned, the return type would be `Future[QueryAssetsResult]`.

**Comment by michaelsbradleyjr** [comment]
Re: all the review comments above, I think some of it is a matter or PR reviews happening in parallel.

@alaibe please let us know if we've confused you in any way, we'll be happy to sort it out!

**Comment by alaibe** [comment]
@michaelsbradleyjr @emizzle thanks both for your review, hopefully I did resolve all the comment and I didn't miss anything :)

I don't think there was any confusion, all the comments seemed to go in the same direction.

File: status/api/opensea.nim [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+  AssetContract* = object
+    address*: Address
+
+  Collection* = object
+    name*: string
+
+  AssetsResult = Result[seq[Asset], string]
+
+proc queryAssets(owner: Address, offset: int): Future[AssetsResult] {.async.} =
```
Comment (emizzle) [inline status/api/opensea.nim:32]:
```suggestion
proc queryAssets(owner: Address, offset, limit: int): Future[AssetsResult] {.async.} =
```

File: status/api/opensea.nim [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  var content = ""
+  let url = fmt("https://api.opensea.io/api/v1/assets?owner={owner}&offset={offset}&limit=50")
```
Comment (emizzle) [inline status/api/opensea.nim:34]:
```suggestion
  let url = fmt("https://api.opensea.io/api/v1/assets?owner={owner}&offset={offset}&limit={limit}")
```

File: status/api/opensea.nim [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  try:
+    let response = await fetch(HttpSessionRef.new(), parseUri(url))
+    content = string.fromBytes(response.data)
+  except CancelledError, HttpError:
+    return AssetsResult.err "Error while fetching assets from opensea"
+
+  let assets = Json.decode(content, AssetsContainer, allowUnknownFields = true).assets
+  return AssetsResult.ok assets
+
+proc getOpenseaAssets*(self: StatusObject, owner: Address): Future[AssetsResult] {.async.} =
```
Comment (emizzle) [inline status/api/opensea.nim:44]:
```suggestion
proc getOpenseaAssets*(self: StatusObject, owner: Address, limit = 50): Future[AssetsResult] {.async.} =
```

File: status/api/opensea.nim [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  var offset = 0
+  var assets: seq[Asset] = @[]
+  while true:
+    let tmpAssets = await queryAssets(owner, offset)
```
Comment (emizzle) [inline status/api/opensea.nim:48]:
```suggestion
    let tmpAssets = await queryAssets(owner, offset, limit)
```

File: status/api/opensea.nim [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    if tmpAssets.isErr: return tmpAssets
+
+    assets = concat(assets, tmpAssets.get())
+    if len(tmpAssets.get()) < 50: break
+
+    offset += 50
```
Comment (emizzle) [inline status/api/opensea.nim:54]:
```suggestion
    if len(tmpAssets.get()) < limit: break

    offset += limit
```

**Review by emizzle — APPROVED** [review]
Great job @alaibe, thanks for taking all the suggestions on board! I have one final suggestion, which is to pass the opensea results limit in to the API call as an optional param (using 50 as the default), just in case clients may want to change the amount of results returned.

It's not a blocker, so the PR could be merged without that change, just thought it may be a simple update and beneficial to end users.

File: status/api/opensea.nim [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+  AssetsResult = Result[seq[Asset], string]
+
+proc queryAssets(owner: Address, offset: int): Future[AssetsResult] {.async.} =
+  var content = ""
+  let url = fmt("https://api.opensea.io/api/v1/assets?owner={owner}&offset={offset}&limit=50")
+  try:
+    let response = await fetch(HttpSessionRef.new(), parseUri(url))
+    content = string.fromBytes(response.data)
+  except CancelledError, HttpError:
```
Comment (emizzle) [inline status/api/opensea.nim:38]: TIL!

**Comment by alaibe** [comment]
No problem for the limit, that would also make it easer if opensea changes its limit params, only 1 place to update!

**Review by michaelsbradleyjr — APPROVED** [review]

**Comment by michaelsbradleyjr** [comment]
Needs a rebase, I'll take care of it now.

**Comment by michaelsbradleyjr** [comment]
I just force-pushed the rebased commit. When all the lights turn green in CI, I intend to merge.

**Comment by michaelsbradleyjr** [comment]
I did a local dev build to try it out the command example client, and discovered a small problem re: the trace in `tui/actions` for `GetAssetsEvent`. I'm fixing it now.

**Comment by michaelsbradleyjr** [comment]
Changes made, including adding a `CommandParameter` to the `help` proc for `GetAssets`.

**Pull request merged**